selfdrive/controls/lib/alertmanager.py

```diff
@@ -44,42 +51,35 @@ def clear_current_alert(self) -> None:
 
   def add_many(self, frame: int, alerts: List[Alert], enabled: bool = True) -> None:
     for alert in alerts:
-      added_alert = copy.copy(alert)
-      added_alert.start_time = frame * DT_CTRL
-
-      # if new alert is higher priority, log it
-      if not len(self.activealerts) or added_alert.alert_priority > self.activealerts[0].alert_priority:
-        cloudlog.event('alert_add', alert_type=added_alert.alert_type, enabled=enabled)
-
-      self.activealerts.append(added_alert)
+      alert_duration = max(alert.duration_sound, alert.duration_hud_alert, alert.duration_text)
+      self.activealerts[alert.alert_type].alert = alert
+      self.activealerts[alert.alert_type].end_frame = frame + int(alert_duration / DT_CTRL)
 
   def process_alerts(self, frame: int, clear_event_type=None) -> None:
-    cur_time = frame * DT_CTRL
-
-    # first get rid of all the expired alerts
-    self.activealerts = [a for a in self.activealerts if a.event_type != clear_event_type and
-                         a.start_time + max(a.duration_sound, a.duration_hud_alert, a.duration_text) > cur_time]
-
-    # sort by priority first and then by start_time
-    self.activealerts.sort(key=lambda k: (k.alert_priority, k.start_time), reverse=True)
-
-    # start with assuming no alerts
-    self.clear_current_alert()
-
-    if len(self.activealerts):
-      current_alert = self.activealerts[0]
-
-      self.alert_type = current_alert.alert_type
-
-      if current_alert.start_time + current_alert.duration_sound > cur_time:
-        self.audible_alert = current_alert.audible_alert
-
-      if current_alert.start_time + current_alert.duration_hud_alert > cur_time:
-        self.visual_alert = current_alert.visual_alert
-
-      if current_alert.start_time + current_alert.duration_text > cur_time:
-        self.alert_text_1 = current_alert.alert_text_1
-        self.alert_text_2 = current_alert.alert_text_2
-        self.alert_status = current_alert.alert_status
-        self.alert_size = current_alert.alert_size
-        self.alert_rate = current_alert.alert_rate
+    current_alert = AlertEntry()
+    for k, v in self.activealerts.items():
+      if v.alert is None:
+        continue
+
+      if v.alert.event_type == clear_event_type:
+        self.activealerts[k].end_frame = -1
+
+      # sort by priority first and then by end_frame
```

Comment:
why sort by end frame? Think it makes more sense to do based on start_frame (which we can store when we create the alert).